### ISSUE TITLE
Integrate Travis-CI

### ISSUE THREAD

**Issue by luzpaz** [andreas-vox/ScribusCTL#2]
## Changelog
- [x] Sign up with [Travis-CI](http://docs.travis-ci.com/user/getting-started/) (free for OSS Projects)
- [x] Ability to trigger Travis CI through any PR or Push
- [x] Figure out where the .travis.yml file will live (perhaps master branch? if and when Travis CI proves to be a good idea; talk to core-devs about putting it in trunk so it's mirrored to github?)
- [x] [Build result notifications](http://docs.travis-ci.com/user/notifications/) through email, [IRC](http://docs.travis-ci.com/user/notifications/#IRC-notification) and much more
- [x] Write a [.travis.yml](http://docs.travis-ci.com/user/build-configuration/) file
  - [x] Linux: @khaledhosny and @QuLogic have created a .yml script for 14.04 
  - [x] OSX: @luzpaz is still developing a [.travis.yml file @ https://github.com/scribusproject/scribus/pull/29) for OSX
  - [ ] Windows: appveyor.yml to be used. See https://github.com/andreas-vox/ScribusCTL/issues/9
## Integrate different tests
- [ ] Every so often trigger a [Coverty](https://scan.coverity.com/travis_ci) scan through Travis CI
- [ ] Follow up with bugtracker issues [Automated "black box" test suite](http://bugs.scribus.net/view.php?id=2231)
- [ ] Integrate Regression Testing at some point (https://github.com/scribusproject/scribus/issues/18)
- [ ] Think of new tests

Mirrored issue at https://github.com/scribusproject/scribus/issues/3
_Copied from original issue: scribusproject/scribus#3_

**Comment by khaledhosny** [comment]
I gave it a try here https://github.com/khaledhosny/ScribusCTL/commits/travis-ci, but Travis CI uses Ubuntu 12.04 which is rather ancient and the build is now breaking on a too old zlib package and I couldn’t find a PPA providing a new one, and I doubt this will be the only breakage.

**Comment by luzpaz** [comment]
@khaledhosny nice work!

> Travis CI uses Ubuntu 12.04 which is rather ancient

This is rather depressing. 
Perhaps we can find another way to get scribus tested in a more up-to-date CI system. Christoph mentioned asking Plinnel and I will do so. As for now..perhaps Travis will just serve to build on OSX which should use a more current OSX for testing (OS X 10.9.5). I have a [homebrew tap formula]([https://github.com/scribusproject/homebrew-scribus/blob/master/scribus.rb]) that I've been working on that compiles from trunk. Usually it works very well but at this very moment it's broken. Fix is in progress.

**Comment by khaledhosny** [comment]
Travis does not accept new projects for Mac OS X building anymore, it was a beta feature or some such. Only Linux is supported, but using that ancient Ubuntu. We can probably get this to work if we can make the Scribus PPA to support 12.04, but this will require backporting a few packages and I don’t have an Ubuntu installation right now.

**Comment by luzpaz** [comment]
@khaledhosny can you provide a link to your failing build ?

**Comment by khaledhosny** [comment]
https://travis-ci.org/khaledhosny/ScribusCTL/builds/63801715

**Comment by luzpaz** [comment]
There is also Appveyor for windows. Thanks to @ebraminio for the heads up on that: https://github.com/khaledhosny/ots/pull/67

**Comment by luzpaz** [comment]
Here is musescore .travis.yml file: https://github.com/musescore/MuseScore/blob/master/.travis.yml

**Comment by khaledhosny** [comment]
Interesting…

**Comment by QuLogic** [comment]
[Halfway there](https://travis-ci.org/QuLogic/ScribusCTL/builds/70136976).

I had to make a few changes: a) patch `ZipPrivate` to work with older zlib (it's pretty simple); b) switch to a newer Qt PPA as it doesn't build with 5.0.z; c) switch to a newer compiler (because again issues compiling). I enabled parallel building because serial builds takes almost 50 minutes (the limit on Travis). I tried [several # of processors](https://travis-ci.org/QuLogic/ScribusCTL/branches) but since Travis has limited CPU and RAM, anything above 8 or so has a flaky speedup.

So I guess the last step is to find a PPA for a newer version of clang if you want to be building with it.

**Comment by QuLogic** [comment]
It doesn't look like [using newer clang works](https://travis-ci.org/QuLogic/ScribusCTL/builds/70145748). It seems to be a bug in OpenSceneGraph, and I've yet to find a PPA for it.

Since OpenSceneGraph is optional, I disabled it for clang _only_, and [now the build passes](https://travis-ci.org/QuLogic/ScribusCTL/builds/70149823).

**Comment by luzpaz** [comment]
Epic @QuLogic !
Yea OpenSceneGraph is very low priority at this point.

**Comment by luzpaz** [comment]
@QuLogic, please make a PR to add that .travis.yml file to the repo

**Comment by QuLogic** [comment]
Clang seems to [get killed](https://travis-ci.org/QuLogic/ScribusCTL/jobs/70159171) pretty often. I'm not sure if it's OOMing or is just buggy. I even tried reducing the parallelism to avoid it (it's only 3 in the mentioned build).

**Comment by luzpaz** [comment]
hmm.. it doesn't happen each time you trigger a build though, right ?

**Comment by QuLogic** [comment]
With NPROC=3, it's failed once and passed once. Let's see how the build in the PR goes.

**Comment by QuLogic** [comment]
Actually, if you count the build in the fork of the main project, then it's passed twice.

**Comment by QuLogic** [comment]
Annnd, it just failed on yours, so it's 50/50. :disappointed:

**Comment by luzpaz** [comment]
What do you mean by: 
`It would be nice to switch to the container based infrastructure where limits are a little higher, but most Scribus dependencies are not whitelisted yet.`
What's a container based infrastructure mean? and how do you know if dependencies are whitelisted or not?

**Comment by QuLogic** [comment]
The [Travis container-based infrastructure](http://docs.travis-ci.com/user/workers/container-based-infrastructure/) is faster (for [booting, officially](http://docs.travis-ci.com/user/ci-environment/#Virtualization-environments), but also for running). But you can't use `sudo` or add apt repositories. Packages can only be installed from [the whitelist](http://docs.travis-ci.com/user/apt/). That's what MuseScore uses as mentioned above.

And it looks like even building [with 2 processors can fail](https://travis-ci.org/QuLogic/ScribusCTL/jobs/70160786). :cry: 

Building with [clang in serial only](https://travis-ci.org/QuLogic/ScribusCTL/jobs/70161746) takes 37 minutes and that leaves a bit of room under the 50 minute limit, but it's kind of annoying.

**Comment by luzpaz** [comment]
I see. Thanks for the explanation. 

Well.. development will be slow with ScribusCTL for the time being. I say this because Andreas told me that he will have time this upcoming weekend to work on it. So we aren't in a hurry because he calls the shots. My opinion: it's ok to not parallelize the build and even give up the gcc build and stick to clang.

**Comment by QuLogic** [comment]
> give up the gcc build and stick to clang.

No, you have it backwards; gcc works in parallel, and clang only works (consistently) in serial.

**Comment by luzpaz** [comment]
@QuLogic @khaledhosny good news, TravisCI is rolling out an invite by request the possibility to build on Ubuntu 14.04. https://twitter.com/travisci/status/619027935857627138

**Comment by QuLogic** [comment]
It looks like that will negate the need for patching OpenSceneGraph and zlib stuff, and also probably won't need a PPA for gcc/g++ and cmake.

**Comment by luzpaz** [comment]
From the Travis CI folks after enabling Ubuntu 14.04 

> You'll need to add `dist: trusty` to each repository's `.travis.yml` file and push that change for builds to run on the Trusty builders.
> 
> Please note that our current Trusty images are more minimal than the precise images and so not all languages and services are installed by default. You may need to do some additional manual setup to get your build environment working. But we're working to get our Trusty images up to parity with the precise ones.
> 
> Thanks

**Comment by QuLogic** [comment]
Does that apply to forks as well, or will I have to work in the PR here?

**Comment by luzpaz** [comment]
Guestimating,  that if it's in your fork, then you need to request to join the beta testing 14.04 program so they can enable your account.

**Comment by luzpaz** [comment]
@QuLogic the devs tell me that we don't need webkit anymore, in regards to the libqt5webkit5-dev in .travis.yml file

**Comment by luzpaz** [comment]
Good news.. multi-os now supported https://twitter.com/travisci/status/629051392217477120
Just shot an email to Travis-CI to enable on these repos. Soon we can trigger builds for OSX as well. Sweet!

**Comment by luzpaz** [comment]
@andreas-vox you can go ahead and copy the SVN copy of .travis.yml in to this repo. It's been updated and also includes MacOSX test build

**Comment by luzpaz** [comment]
@andreas-vox I know this is semi low priority, but we can now test PRs through travis again after the script was fixed. Feel free to cherry pick 
r20494 (Quick reference:  7c9123c741eb3d6e81a3e3ff8984985cb86297e4 [link](https://github.com/scribusproject/scribus/commit/7c9123c741eb3d6e81a3e3ff8984985cb86297e4) )
and 
r20505 (Quick reference:  cc574402b3a65d2da06d0068ba7b0baddaa7ffa4 [link](https://github.com/scribusproject/scribus/commit/cc574402b3a65d2da06d0068ba7b0baddaa7ffa4) )

**Comment by luzpaz** [comment]
Closing this issue in favor of https://github.com/scribusproject/scribus/issues/3
